### PR TITLE
Add by_domain_doc_type_date view

### PR DESCRIPTION
**Pull request by esoergel**
@dannyroberts @millerdev cc: @nickpell
`cherry-pick`ing from Danny's PR here: https://github.com/dimagi/commcare-hq/pull/8390
https://github.com/dimagi/fluff/pull/112 should be merged at the same time as this.
I'm hoping to get this in after Friday's deploy to kick off a long reindex.

### CODE REVIEW

File: corehq/couchapps/by_domain_doc_type/views/view/map.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+function (doc) {
```
Comment (dimagimon) [inline corehq/couchapps/by_domain_doc_type/views/view/map.js]: Missing name in function declaration.

File: corehq/couchapps/__init__.py [hunk 2]
```diff
@@ -6,4 +6,6 @@
     'users_extra': (settings.USERS_GROUPS_DB, settings.NEW_USERS_GROUPS_DB),
     'noneulized_users': (settings.USERS_GROUPS_DB, settings.NEW_USERS_GROUPS_DB),
     'all_docs': (None, settings.NEW_USERS_GROUPS_DB, settings.NEW_FIXTURES_DB),
+    # TODO should we make an '_all' option or something?
+    'by_domain_doc_type': (None, settings.NEW_USERS_GROUPS_DB, settings.NEW_FIXTURES_DB),
```
Comment (esoergel) [inline corehq/couchapps/__init__.py]: I'm thinking about how to future-proof against subsequent migrations - seems like `all_docs` and `by_domain_doc_type` should exist in ~every db.

**Comment by czue** [comment]
build fail real. I think the appropriate thing to do is to keep the mixin and change the view name.

```
======================================================================
ERROR: test_get (corehq.apps.cachehq.tests.TestCachedCouchDocumentMixin)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/mock.py", line 1224, in patched
    return func(*args, **keywargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/cachehq/tests.py", line 28, in test_get
    blog_post.save()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/cachehq/mixins.py", line 17, in save
    self.clear_caches()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/cachehq/mixins.py", line 46, in clear_caches
    super(QuickCachedDocumentMixin, self).clear_caches()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/cachehq/mixins.py", line 11, in clear_caches
    invalidate_document(self)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/cachehq/invalidate.py", line 16, in invalidate_document
    get_cache_pillow().change_trigger({
  File "submodules/dimagi-utils-src/dimagi/utils/decorators/memoized.py", line 11, in _inner
    return m(*args, **kwargs)
  File "submodules/dimagi-utils-src/dimagi/utils/decorators/memoized.py", line 127, in __call__
    value = self.func( *args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/cachehq/invalidate.py", line 7, in get_cache_pillow
    return cacheinvalidate.CacheInvalidatePillow()
  File "/home/travis/build/dimagi/commcare-hq/corehq/pillows/cacheinvalidate.py", line 21, in __init__
    self.gen_caches = set(GenerationCache.doc_type_generation_map().values())
  File "submodules/dimagi-utils-src/dimagi/utils/couch/cache/cache_core/gen.py", line 32, in doc_type_generation_map
    for gen_model in GenerationCache._get_generational_caches():
  File "submodules/dimagi-utils-src/dimagi/utils/couch/cache/cache_core/gen.py", line 18, in _get_generational_caches
    GenerationCache._generate_caches()
  File "submodules/dimagi-utils-src/dimagi/utils/couch/cache/cache_core/gen.py", line 44, in _generate_caches
    gen_model = getattr(mod, cache_class_name)
AttributeError: 'module' object has no attribute 'CommtrackConfigGenerationCache'
```

File: corehq/couchapps/by_domain_doc_type_date/views/view/map.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+function (doc) {
```
Comment (dimagimon) [inline corehq/couchapps/by_domain_doc_type_date/views/view/map.js]: Missing name in function declaration.

File: corehq/couchapps/by_domain_doc_type_date/views/view/map.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            case "CommCareCase-Deleted":
+                return doc.opened_on;
+            case "XFormInstance":
+            case "XFormInstance-Deleted":
+            case "XFormError":
+            case "XFormDuplicate":
+            case "XFormDeprecated":
+            case "XFormArchived":
+            case "SubmissionErrorLog":
+                return doc.received_on
```
Comment (dimagimon) [inline corehq/couchapps/by_domain_doc_type_date/views/view/map.js]: Missing semicolon.

**Comment by dannyroberts** [comment]
Two different test failures:
```
======================================================================
ERROR: setUpClass (corehq.apps.locations.tests.test_location_expressions.TestLocationTypeExpression)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/testrunner.py", line 139, in new_setUpClass
    result = original_setUpClass(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/testrunner.py", line 139, in new_setUpClass
    result = original_setUpClass(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/locations/tests/test_location_expressions.py", line 19, in setUpClass
    cls.location_type.save()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/locations/models.py", line 108, in save
    self._populate_stock_levels()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/locations/models.py", line 95, in _populate_stock_levels
    or (not Domain.get_by_name(self.domain).commtrack_enabled)
AttributeError: 'NoneType' object has no attribute 'commtrack_enabled'
```
and
```
======================================================================
ERROR: test_device_report_submissions_ignored (corehq.apps.commtrack.tests.test_xml.BugSubmissionsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/commtrack/tests/test_xml.py", line 173, in setUp
    super(CommTrackSubmissionTest, self).setUp()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/commtrack/tests/util.py", line 159, in setUp
    self.ct_settings.save()
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/cachehq/mixins.py", line 15, in save
    super(_InvalidateCacheMixin, self).save(**params)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/couchdbkit/schema/base.py", line 134, in save
    db.save_doc(doc, **params)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/couchdbkit/client.py", line 515, in save_doc
    **params).json_body
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/restkit/resource.py", line 153, in put
    headers=headers, params_dict=params_dict, **params)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/couchdbkit/resource.py", line 140, in request
    response=e.response)
ResourceConflict: Document update conflict.
```

**Comment by esoergel** [comment]
This fixed the tests: https://github.com/dimagi/commcare-hq/commit/d51c0baf13ba5c4e7cbd645876bb509e15ceb0f8
When I switched `CommtrackConfig` to `quickcache`, it wasn't being invalidated on `Domain.delete()` (which uses `save_docs` and doesn't call the model methods), so subsequent tests using the same test domain name `"test-domain"` were seeing that when they shouldn't have.

**Pull request merged**